File: rag/chunker.py
import re
from typing import List
from .types import Chunk, ChunkingStrategy, DocumentItem

def estimate_token_count(text: str) -> int:
    """
    Estimates token count for English text (~1.3 tokens per word)
    """
    if not text or not text.strip():
        return 0
    words = len(text.strip().split())
    return max(1, int(round(words * 1.3)))

def chunk_document(
    doc: DocumentItem,
    chunk_size: int = 300,
    chunk_overlap: int = 50,
    strategy: ChunkingStrategy = 'fixed-size'
) -> List[Chunk]:
    """
    Splits document text content into semantic Chunk objects.
    """
    content = doc.content
    if not content or not content.strip():
        return []

    chunks: List[Chunk] = []

    if strategy == 'paragraph':
        paragraphs = [p.strip() for p in re.split(r'\n\s*\n', content) if p.strip()]
        current_text = ""
        start_idx = 0

        for p in paragraphs:
            if len(current_text + "\n\n" + p) > chunk_size and len(current_text) > 0:
                chunks.append(_create_chunk(doc, len(chunks), current_text.strip(), start_idx))
                start_idx += len(current_text)
                current_text = p
            else:
                if not current_text:
                    start_idx = content.find(p)
                current_text = f"{current_text}\n\n{p}" if current_text else p

        if current_text.strip():
            chunks.append(_create_chunk(doc, len(chunks), current_text.strip(), start_idx))

    elif strategy == 'sentence':
        # Sentence splitting pattern
        sentences = [s.strip() for s in re.split(r'(?<=[.!?])\s+', content) if s.strip()]
        current_text = ""
        start_idx = 0

        for s in sentences:
            if len(current_text + " " + s) > chunk_size and len(current_text) > 0:
                chunks.append(_create_chunk(doc, len(chunks), current_text.strip(), start_idx))
                start_idx += len(current_text)

                if chunk_overlap > 0 and len(current_text) > chunk_overlap:
                    overlap_text = current_text[-chunk_overlap:]
                    current_text = overlap_text + " " + s
                else:
                    current_text = s
            else:
                current_text = f"{current_text} {s}" if current_text else s

        if current_text.strip():
            chunks.append(_create_chunk(doc, len(chunks), current_text.strip(), start_idx))

    else:
        # Fixed character size with overlap
        i = 0
        length = len(content)
        chunk_idx = 0

        while i < length:
            end = min(i + chunk_size, length)
            chunk_text = content[i:end]

            # Avoid splitting words mid-character if possible
            if end < length:
                last_space = chunk_text.rfind(' ')
                if last_space > int(chunk_size * 0.6):
                    chunk_text = chunk_text[:last_space]

            trimmed = chunk_text.strip()
            if trimmed:
                chunks.append(_create_chunk(doc, chunk_idx, trimmed, i))
                chunk_idx += 1

            actual_step = len(chunk_text) - (chunk_overlap if chunk_overlap < len(chunk_text) else 0)
            i += max(1, actual_step)

    return chunks

def _create_chunk(doc: DocumentItem, index: int, text: str, start_char_idx: int) -> Chunk:
    return Chunk(
        id=f"{doc.id}-chunk-{index}",
        documentId=doc.id,
        documentName=doc.name,
        chunkIndex=index,
        text=text,
        tokenCount=estimate_token_count(text),
        characterCount=len(text),
        startCharIndex=start_char_idx,
        endCharIndex=start_char_idx + len(text),
        tags=doc.tags or []
    )

File: src/services/chunker.ts
import type { Chunk, ChunkingStrategy, DocumentItem } from '../types/rag';

export function estimateTokenCount(text: string): number {
  if (!text) return 0;
  const words = text.trim().split(/\s+/).filter(Boolean).length;
  return Math.max(1, Math.round(words * 1.3));
}

export function chunkDocument(
  doc: DocumentItem,
  chunkSize: number = 300,
  chunkOverlap: number = 50,
  strategy: ChunkingStrategy = 'fixed-size'
): Chunk[] {
  const content = doc.content;
  if (!content || content.trim().length === 0) return [];

  const chunks: Chunk[] = [];

  if (strategy === 'paragraph') {
    const paragraphs = content.split(/\n\s*\n/).filter((p) => p.trim().length > 0);
    let currentChunkText = '';
    let startIdx = 0;

    paragraphs.forEach((p) => {
      if ((currentChunkText + '\n\n' + p).length > chunkSize && currentChunkText.length > 0) {
        chunks.push(createChunkObj(doc, chunks.length, currentChunkText.trim(), startIdx));
        startIdx += currentChunkText.length;
        currentChunkText = p;
      } else {
        if (!currentChunkText) startIdx = content.indexOf(p);
        currentChunkText = currentChunkText ? `${currentChunkText}\n\n${p}` : p;
      }
    });

    if (currentChunkText.trim().length > 0) {
      chunks.push(createChunkObj(doc, chunks.length, currentChunkText.trim(), startIdx));
    }
  } else if (strategy === 'sentence') {
    const sentences = content.match(/[^.!?]+[.!?]+(\s+|$)|[^.!?]+$/g) || [content];
    let currentChunkText = '';
    let startIdx = 0;

    sentences.forEach((s) => {
      if ((currentChunkText + ' ' + s).length > chunkSize && currentChunkText.length > 0) {
        chunks.push(createChunkObj(doc, chunks.length, currentChunkText.trim(), startIdx));
        startIdx += currentChunkText.length;

        if (chunkOverlap > 0 && currentChunkText.length > chunkOverlap) {
          const overlapText = currentChunkText.slice(-chunkOverlap);
          currentChunkText = overlapText + ' ' + s;
        } else {
          currentChunkText = s;
        }
      } else {
        currentChunkText = currentChunkText ? `${currentChunkText} ${s}` : s;
      }
    });

    if (currentChunkText.trim().length > 0) {
      chunks.push(createChunkObj(doc, chunks.length, currentChunkText.trim(), startIdx));
    }
  } else {
    let i = 0;
    const len = content.length;
    let chunkIndex = 0;

    while (i < len) {
      const end = Math.min(i + chunkSize, len);
      let chunkText = content.slice(i, end);

      if (end < len) {
        const lastSpace = chunkText.lastIndexOf(' ');
        if (lastSpace > chunkSize * 0.6) {
          chunkText = chunkText.slice(0, lastSpace);
        }
      }

      const trimmed = chunkText.trim();
      if (trimmed.length > 0) {
        chunks.push(createChunkObj(doc, chunkIndex, trimmed, i));
        chunkIndex++;
      }

      const actualStep = chunkText.length - (chunkOverlap < chunkText.length ? chunkOverlap : 0);
      i += Math.max(1, actualStep);
    }
  }

  return chunks;
}

function createChunkObj(doc: DocumentItem, index: number, text: string, startCharIndex: number): Chunk {
  return {
    id: `${doc.id}-chunk-${index}`,
    documentId: doc.id,
    documentName: doc.name,
    chunkIndex: index,
    text,
    tokenCount: estimateTokenCount(text),
    characterCount: text.length,
    startCharIndex,
    endCharIndex: startCharIndex + text.length,
    tags: doc.tags || [],
  };
}
